### PR TITLE
DOC: Fix sphinx warnings in c-info.beyond-basics.rst

### PR DESCRIPTION
**Pull request by cmarmo**
This pull request partially addresses #13114
In particular it fixes the sphinx warnings in [c-info.beyond-basics.rst](https://numpy.org/doc/stable/user/c-info.beyond-basics.html#beyond-the-basics).
```
/home/circleci/repo/doc/source/user/c-info.beyond-basics.rst:162: WARNING: c:identifier reference target not found: PyArrayMultiObject
/home/circleci/repo/doc/source/user/c-info.beyond-basics.rst:162: WARNING: c:func reference target not found: PyArray_MultiIter_SIZE
/home/circleci/repo/doc/source/user/c-info.beyond-basics.rst:264: WARNING: c:func reference target not found: PyArray_TypeNumFromName
/home/circleci/repo/doc/source/user/c-info.beyond-basics.rst:466: WARNING: c:func reference target not found: PyArray_FinalizeFunc
```
Hope my guesses are the right choices.

### CODE REVIEW

File: doc/source/reference/c-api/array.rst [hunk 1]
```diff
@@ -1278,6 +1278,14 @@ New data types
...
     return the previously-assigned type-number.
 
+.. c:function:: int PyArray_TypeNumFromName( \
+        char const *str)
+
+   Given a string return the type-number for the data-type with that string as
+   the type-object name.
+   Returns NPY_NOTYPE without setting an error if no type can be found.
+   Only works for user-defined data-types.
+
```
Comment (mattip) [inline doc/source/reference/c-api/array.rst:1288]: I think this belongs at the end of the section, which would order these functions alphabetically and also make the helper function used after registering the user-defined dtype come after all the registering functions. Also may be worth changing the heading from `New data types` to `User-defined data types`.

File: doc/source/reference/c-api/array.rst [hunk 2]
```diff
@@ -2555,6 +2563,10 @@ Broadcasting (multi-iterators)
     through all of the elements (of the broadcasted result), otherwise
     it evaluates FALSE.
 
+.. c:function:: int PyArray_MultiIter_SIZE(PyObject* multi)
+
+    Returne the size of the multi-iterator object.
```
Comment (mattip) [inline doc/source/reference/c-api/array.rst:2568]:
```suggestion
    Returns the size of the multi-iterator object.
```

File: doc/source/user/c-info.beyond-basics.rst [hunk 3]
```diff
@@ -472,7 +472,7 @@ The __array_finalize\__ method
    members are filled in. Finally, the :obj:`~numpy.class.__array_finalize__`
    attribute is looked-up in the object dictionary. If it is present
    and not None, then it can be either a CObject containing a pointer
-   to a :c:func:`PyArray_FinalizeFunc` or it can be a method taking a
+   to a ``PyArray_FinalizeFunc`` or it can be a method taking a
```
Comment (mattip) [inline doc/source/user/c-info.beyond-basics.rst:475]: I think we should leave this as a hanging reference to remind us to document this function. Unfortunately it is not even documented in the [public header file](https://github.com/numpy/numpy/blob/df6d2600c51502e1877aac563658d0616a75c5e5/numpy/core/include/numpy/ndarraytypes.h#L793) where it is declared, and I don't see any example use of it in tests.

File: numpy/core/include/numpy/ndarraytypes.h [hunk 4]
```diff
@@ -1308,6 +1308,8 @@ typedef struct {
 #define PyArray_MultiIter_NOTDONE(multi)                \
         (_PyMIT(multi)->index < _PyMIT(multi)->size)
 
+#define PyArray_MultiIter_SIZE(multi)                \
+        _PyMIT(multi)->size
```
Comment (mattip) [inline numpy/core/include/numpy/ndarraytypes.h:1312]: If you really want to expose a new macro, it should be in a separate PR, not in a documentation PR.

File: doc/source/reference/c-api/array.rst (hunk 2 above)
Comment (mattip) [reply doc/source/reference/c-api/array.rst:2568]: Why are you adding this macro?

**Comment by mattip** [comment]
Something is weird with `PyArray_MultiIter_SIZE`. I don't understand why it [is documented](https://21173-908607-gh.circle-artifacts.com/0/doc/build/html/user/c-info.beyond-basics.html?highlight=pyarray_multiiter_size#broadcasting-over-multiple-arrays) but never defined. Did it get removed in a refactor?

**Comment by mattip** [comment]
Even in the [initial version of the NumPy docs](03582a3dbd) from 2008, the documentation mentions the macro but it is not in NumPy.

**Comment by cmarmo** [comment]
Thanks @mattip for you comments.
Indeed, I was not sure about `PyArray_MultiIter_SIZE` and `PyArray_FinalizeFunc`.
To clarify, you are suggesting
- to completely remove `PyArray_MultiIter_SIZE` from the documentation
- to add  `PyArray_FinalizeFunc` in, let's say, `doc/source/reference/c-api/array.rst` (or wherever this is appropriate... )

Am I understanding correctly? Thanks.

**Comment by mattip** [comment]
> to completely remove `PyArray_MultiIter_SIZE` from the documentation

Then we would need to redo that example that uses it. Perhaps leave it out of this PR for now

> to add `PyArray_FinalizeFunc` ...

Yes, but what should that documentation be?

**Comment by cmarmo** [comment]
> > to completely remove `PyArray_MultiIter_SIZE` from the documentation
> 
> Then we would need to redo that example that uses it. Perhaps leave it out of this PR for now

Ok, if you don't mind I will open an issue as a follow-up, to track this never defined `PyArray_MultiIter_SIZE` 

> > to add `PyArray_FinalizeFunc` ...
> 
> Yes, but what should that documentation be?

Well, I have tried something... it seems a bit of a circular definition... but at least there is something that could be improved in the future and fix the warning. Let me know if this is acceptable. Thanks.

**Review by mattip — APPROVED** [review]
Looks good to me except for one small nit.

File: doc/source/reference/c-api/array.rst [hunk 5]
```diff
@@ -151,6 +151,16 @@ and its sub-types).
...
 
+.. c:function:: int PyArray_FinalizeFunc(PyArrayObject* arr, PyObject* obj)
+
+    The function pointed to by the CObject
+    :obj:`~numpy.class.__array_finalize__`.
+    The first argument is the newly created sub-type. The second argument
+    (if not NULL) is the "parent" array (if the array was created using
+    slicing or some other operation where a clearly-distinguishable parent
+    is present). This routine can do anything it wants to. It should
+    return a -1 on error and 0 otherwise.
```
[the remark this answers is not in the document]
Comment (cmarmo) [reply doc/source/reference/c-api/array.rst:162]: :innocent:

**Comment by mattip** [comment]
Only [296 warnings](https://app.circleci.com/pipelines/github/numpy/numpy/9096/workflows/0d3aa23c-3f93-4822-b5a0-d176feddca23/jobs/21194) left. It seems like many of them are due to the use of `size` in the random module like here https://github.com/numpy/numpy/blob/df6d2600c51502e1877aac563658d0616a75c5e5/numpy/random/_generator.pyx#L149-L152

**Comment by mattip** [comment]
Thanks @cmarmo
